riocli/device/model.py

```diff
@@ -48,33 +44,84 @@ def find_object(self, client: Client) -> bool:
 
     def create_object(self, client: Client, **kwargs) -> v1Device:
         if not self.spec.get('virtual', {}).get('enabled', False):
-            device = client.create_device(self.to_v1())
-            return device
+            try:
+                device = client.create_device(self.to_v1())
+                return device
+            except Exception as e:
+                raise e
 
-        hwil_response = create_hwil_device(self.spec, self.metadata)
         try:
-            device_uuid = find_device_guid(client, self.metadata.name)
-            device = client.get_device(device_uuid)
-        except DeviceNotFound:
-            update_device_labels(self.metadata, hwil_response)
+            # Create the HWIL device.
+            hwil_response = create_hwil_device(self.spec, self.metadata)
+
+            # Generate labels to store HWIL metadata in rapyuta.io device.
+            l = make_device_labels_from_hwil_device(hwil_response)
+            self.metadata.get('labels', {}).update(l)
+
+            # Create the rapyuta.io device.
             device = client.create_device(self.to_v1())
+
+            # On-board the HWIL device using the onboard script.
+            onboard_script = device.onboard_script()
+            onboard_command = onboard_script.full_command()
+            execute_onboard_command(hwil_response.id, onboard_command)
+
+            return device
         except Exception as e:
             raise e
 
-        onboard_script = device.onboard_script()
-        onboard_command = onboard_script.full_command()
+    def update_object(self, client: Client, obj: typing.Any) -> typing.Any:
+        virtual = self.spec.get('virtual', {}).get('enabled', False)
+
+        if not virtual:
+            return
+
+        device_uuid = find_device_guid(client, self.metadata.name)
+        device = client.get_device(device_uuid)
+
+        # Nothing to do if the device is already online or initializing.
+        if device['status'] in ('ONLINE', 'INITIALIZING'):
+            return
+
+        device_labels = device.get('labels', {})
+        # Convert list to dict for easy access.
+        device_labels = {l['key']: l for l in device_labels}
+
+        # Otherwise, re-onboard the device.
         try:
+            hwil_response = create_hwil_device(self.spec, self.metadata)
```

Comment:
suggestion: Consider checking if the device from the metadata already exists. In the creation, the failure might have happened after the device was created (maybe during the onboarding).